### ISSUE TITLE
How is Peak Memory per GPU lower for Full finetune compared to LoRA?

### ISSUE THREAD

**Issue by NirmalManoj** [pytorch/torchtune#794]
<img width="640" alt="image" src="https://github.com/pytorch/torchtune/assets/18501167/bcc98f22-ffd6-48cb-a41c-f60e0bcafd76">

**Comment by kartikayk** [comment]
Great observation! So the numbers quoted in the README are related to the default configs.

Our single device full-fine-tune recipe has a few optimizations that the default LoRA config doesn't enable. Eg: we have `optimizer_in_bwd=True` which fuses the optimizer step with backward and reduces the memory footprint associated with gradients (see https://pytorch.org/tutorials/intermediate/optimizer_step_in_backward_tutorial.html for more detail). We also make use of the PagedAdamW from bitsandbytes in the full-finetune recipe compared to standard AdamW in LoRA.

There aren't any technical reasons that stops us from enabling these for LoRA. But full-finetune definitely needed more memory-optimization love to get up and running on single GPU with 24GB hence these defaults. We'll have a detailed tutorial on this topic coming out soon :)

Note: There a small gotcha here - you can't use optimizer_in_bwd with gradient accumulation (no gradients to accumulate!) and so that's something to keep in mind.

**Comment by NirmalManoj** [comment]
Thanks for explaining, @kartikayk. I think "gotcha" needs to be highlighted!
